Fix brake function, add softStart (accelerate from zero)

ronso0 opened the pull request:
Continuing to fix https://bugs.launchpad.net/mixxx/+bug/1571442
"Brake effect initial speed and factor"Initial speed was fixed in #1072.
This PR makes engine.brake() actually take the deceleration factor into account by initializing AlphaBetaFilter with custom beta (How was that done originally at the time the docuwiki was written?).
Additionally, one can now call script.brake(..., factor) in controller script.

Commit:
Make brake() consider deceleration factor

Commit:
Update controllerengine.cpp

Commit:
Update common-controller-scripts.js

Commit:
default factor = 1

Be-ing commented on res/controllers/common-controller-scripts.js:
```diff
@@ -315,6 +315,10 @@ script.spinback = function(channel, control, value, status, group) {
    -------- ------------------------------------------------------ */
 script.brake = function(channel, control, value, status, group, factor) {
+    // set default factor
+    if (!factor) {
```
This will not work as you expect because of JavaScript's type coercion. Use `factor === undefined`

Commit:
Update common-controller-scripts.js

ronso0 commented:
Should I remove the whole m_rampFactor from brake() function?
I don't see how it affects the brake effect at all.

ronso0 commented:
..forget about that.
It can be used when ramping to 1 like with a softStart function to accelerate to 1, see
https://bugs.launchpad.net/mixxx/+bug/1643720
Although I would probably create a whole new function instead of making brake() more complex.

Commit:
remove unused 'delay' from function description

ronso0 commented:
Please don't merge yet, I'll extend this PR with further clean-up and softStart() function.

Commit:
Add files via upload

Commit:
adapt to updated brake() & softStart function

removed calculation of current deck rate, this is done in brake/softStart function when no/default rate values are passed.

ronso0 commented:
...and updated common-controller-scripts.js accordingly

ronso0 commented:
*bump*
I'm done, this can be merged if it's okay. Travis errors aren't related to this code, as far as I can see.
Is clear how softStart() works or should I add comments?

Be-ing commented on res/controllers/common-controller-scripts.js:
```diff
@@ -311,18 +311,32 @@ script.spinback = function(channel, control, value, status, group) {
+   Output:  none
+   -------- ------------------------------------------------------ */
+script.softStart = function(channel, control, value, status, group, factor) {
+	if (factor == undefined) {
```
use `===`

Be-ing commented on res/controllers/common-controller-scripts.js:
```diff
@@ -311,18 +311,32 @@ script.spinback = function(channel, control, value, status, group) {
-    // disable on note-off or zero value note/cc, use default decay rate '1', consider playback speed
-    engine.brake(parseInt(group.substring(8,9)), ((status & 0xF0) != 0x80 && value > 0), factor, currentRate);
+    // disable on note-off or zero value note/cc, use default decay rate '1'
+    engine.brake(parseInt(group.substring(8,9)), ((status & 0xF0) != 0x80 && value > 0), factor);
```
use `!==`

Be-ing commented on res/controllers/common-controller-scripts.js:
```diff
@@ -311,18 +311,32 @@ script.spinback = function(channel, control, value, status, group) {
+		factor = 1;
+	}
+    // disable on note-off or zero value note/cc, use default increase rate '1'
+    engine.softStart(parseInt(group.substring(8,9)), ((status & 0xF0) != 0x80 && value > 0), 1);
```
use `!==`

Be-ing commented:
Noticed some small JavaScript nitpicks glancing at the code. I'll give this a try soon.

Commit:
changed !=/== to !==/===

ronso0 commented:
Okay, thank you! updated
Those === and !== are actually copy/paste from existing code.

softStart() is in general like brake() but has different rampTo[] value and turns on 'play'.
scratchProcess() now considers m_softStartActive[] and uses rampTo[] value as target rate when to turn off scratching.

Be-ing reviewed (changes requested):
I tested this and compared it to current master. I confirm that it fixes [the bug](https://bugs.launchpad.net/mixxx/+bug/1571442).

It would be nice if the deck was stopped when there is practically no audio coming through. I tried using script.brake() with really low factors (< .3). I held down the button for a long time after sound stopped coming through, but the deck did not actually stop; when I let go of the button, it started playing at its normal speed again. If you don't want to solve this (or don't want to do it right now), that's fine. I think this PR could be merged without that.

Commit:
engine.softStart(..., factor);

ronso0 commented on res/controllers/common-controller-scripts.js:
```diff
@@ -303,22 +303,40 @@ script.midiPitch = function (LSB, MSB, status) {
+		factor = 1;
+	}
+    // disable on note-off or zero value note/cc, use default increase rate '1'
+    engine.softStart(parseInt(group.substring(8,9)), ((status & 0xF0) !== 0x80 && value > 0), 1);
```
Indeed. Fixed.

ronso0 commented:
> It would be nice if the deck was stopped when there is practically no audio coming through. I tried using script.brake() with really low factors (< .3)

Yeah, that would be nice! So, there had to be a minimum rate, something like setting rampTo to 0.001 for brake(). I'll experiment with what sounds reasonable.
Now that you mention it, I guess the same should be applied to softStart() when it's triggered with really low factors. I'm confident I can finish this until the weekend, then it can be merged.

> when I let go of the button, it started playing at its normal speed again.

I hope you don't mind that, it was intended for original brake/spinback. I'm quite sure you noticed that you can map start and stop of brake() or softStart() to different buttons, and then sending just NoteOn (NoteOff respetively).
I can add that explicit info to the wiki to clarify the usage of those two functions. That would also solve this bug I guess https://bugs.launchpad.net/mixxx/+bug/1090780

Be-ing commented:
> I hope you don't mind that, it was intended for original brake/spinback. I'm quite sure you noticed that you can map start and stop of brake() or softStart() to different buttons, and then sending just NoteOn (NoteOff respetively).

Yeah, they could be mapped that way. They could also be mapped to replace a normal play button.

Be-ing commented:
@Pegasus-RPG any thoughts on this?

ronso0 commented:
>>It would be nice if the deck was stopped when there is practically no audio coming through. I tried using script.brake() with really low factors (< .3)

>Yeah, that would be nice! So, there had to be a minimum rate, something like setting rampTo to 0.001 for brake(). I'll experiment with what sounds reasonable.

I tried to set the minimum rate via rampTo but that would make the unwanted effect even worse.
I'll check if setting a minimum rate in scratchProcess (for brake/spinback only)
Will be back in a few days, I guess.

Be-ing commented:
I think this should be merged now. Improvements for handling very low factors can come in a future PR.

ronso0 commented:
Yup! As soon as this is merged, I'll complete the wiki, including a hint for the low factors.

radusuciu commented on res/controllers/common-controller-scripts.js:
```diff
@@ -303,22 +303,40 @@ script.midiPitch = function (LSB, MSB, status) {
+		factor = 1;
+	}
+    // disable on note-off or zero value note/cc, use default increase rate '1'
+    engine.softStart(parseInt(group.substring(8,9)), ((status & 0xF0) !== 0x80 && value > 0), factor);
```
Indent issue?

ronso0 replied:
Indeed, those 3 lines above had tabs. Fixed.

Commit:
Update common-controller-scripts.js

daschuer commented:
Thank your for the PR and the review!

Pegasus-RPG commented:
Especially since features like this are nuanced and carefully determined, PLEASE add tests to ensure that future changes don't adversely affect the behavior.

ronso0 commented:
Cool, thanks for merging! I'll add info to the wiki.
@Pegasus-RPG Unfortunately I have no idea how to write tests. Though, I'm eager to learn, so it would be great if someone else writes those test and I follow that.
There's no test for brake function so far. Because that it's not necessary/possible or just that nobody wrote one yet?

Be-ing commented:
There's no test because no one wrote one yet. There is a [quick intro](http://mixxx.org/wiki/doku.php/unit_tests) to writing tests on the wiki.
